Assets/Shopify/UIToolkit/Theme Controllers/ThemeControllerBase.cs

```diff
@@ -1,44 +1,72 @@
 ﻿namespace Shopify.UIToolkit {
     using UnityEngine;
     using Shopify.Unity;
+    using Shopify.Unity.SDK;
 
     public abstract class ThemeControllerBase : MonoBehaviour {
         /// <summary>
         /// The Shop Domain to connect to, in the form of "myshop.myshopify.com"
         /// </summary>
-        public string ShopDomain;
+        public string ShopDomain {
+            get { return _shopDomain; }
+            set {
+                _shopDomain = value;
+                InvalidateClient();
+            }
+        }
+        [SerializeField] private string _shopDomain;
 
         /// <summary>
         /// The Storefront Access Token used to authenticate with the connected shop
         /// </summary>
-        public string AccessToken;
+        public string AccessToken {
+            get { return _accessToken; }
+            set {
+                _accessToken = value;
+                InvalidateClient();
+            }
+        }
+        [SerializeField] private string _accessToken;
+
+        /// <summary>
+        /// What loader provider this controller uses to create the loader.
+        /// You can change this if you require a special kind of request loader
+        /// to be used instead of the default UnityLoader.
+        /// </summary>
+        public ILoaderProvider LoaderProvider {
+            get { return _loaderProvider; }
+            set {
+                _loaderProvider = value;
+                InvalidateClient();
+            }
+        }
+        private ILoaderProvider _loaderProvider = new UnityLoaderProvider();
 
         /// <summary>
         /// The client that the controller is using to make requests against the SDK.
         /// </summary>
         /// <returns></returns>
         public ShopifyClient Client {
             get {
-                if (_cachedClient != null &&
-                    _cachedClient.AccessToken == AccessToken && 
-                    _cachedClient.Domain == ShopDomain) {
-                        return _cachedClient;
+                if (_cachedClient == null) {
```

Comment:
This is kind of cool because it defers the creation until someone tries to use the Client but I was wondering, when invalidating the client, could we not just regenerate it then?

Reply:
yea we could for sure... can't think of a reason why one approach is better than the other though